src/opossum_lib/opossum_file.py

```diff
@@ -60,32 +61,74 @@ class Metadata:
     buildDate: Optional[str] = None
 
 
+class ResourceType(Enum):
+    FILE = auto()
+    FOLDER = auto()
+    TOP_LEVEL = auto()
+    OTHER = auto()
+
+
 @dataclass(frozen=True)
 class Resource:
+    type: ResourceType
     children: Dict[str, Resource] = field(default_factory=dict)
 
-    def add_path(self, path: List[str]) -> None:
+    def add_path(self, path: List[str], type_of_last_element: ResourceType) -> None:
         if len(path) == 0:
             return
         first, rest = path[0], path[1:]
+        if self.leaf_element_exists_but_resource_type_differs(
+            first, rest, type_of_last_element
+        ):
+            raise TypeError(
+                "Couldn't add path to resource: ResourceType of leaf elements with"
+                "the same path differ."
+            )
         if first not in self.children:
-            self.children[first] = Resource()
-        self.children[first].add_path(rest)
+            if len(rest) != 0:
+                self.children[first] = Resource(type=ResourceType.FOLDER)
+            else:
+                self.children[first] = Resource(type=type_of_last_element)
+
+        self.children[first].add_path(rest, type_of_last_element)
+
+    def leaf_element_exists_but_resource_type_differs(
+        self, first: str, rest: List[str], type_of_last_element: ResourceType
```

Comment:
not a fan of the variable names first and rest; its fine inside `def add_path` where the splitting occurs but now that its passed to another function, all context behind the names is lost.